design/1.0.md

```diff
@@ -0,0 +1,84 @@
+Values should be specifiable via any or all of:
+
+* Type / Field: Package path and type name, optional field name
+* Field tags
+* Scope: local variable, free variable, or global
+* Context: within a specific package or function call.  See CallSpecifier below.
+* Is a reference
+* Is "reference-like," e.g., slices, maps, chans.
+
+#### CallSpecifier
+A CallSpecifier marks an `*ssa.Call`, e.g., as a sink or sanitizer.
+Calls should be specifiable via any or all of:
+
+* Symbol: package path, function name, optional receiver name.
+* Context: A CallSpecifier indicating  
+* Based on arguments (ValueSpecifier)
+* Based on return values (ValueSpecifier)
+
+## Inference
+
+Inference represents the scope of what a user does *not* need to configure.
+For instance, we can currently detect getter methods and mark as propagating
+a marked field without requiring any explicit configuration.
+
+If a source type is aliased, the alias is properly identified as equivalent as the aliased source type.
+
+# Beyond a 1.0
+
+The following seem like significant improvements but seem larger in scope than what constitutes a stable 1.0.
+These items are open to interpretation and warrant discussion.
+
+## Improved Inference
+
+* Detect sanitization via zero-value.
+    * Explicit sanitization of `type Source struct{ data string }` is easily done via `s.data = ""`.
+    * Alternatively, a user may specify redaction placeholders, e.g. `"REDACTED"` rather than require a zero value.
+* A type defined from a source type (e.g., `type Foo core.Source`) should also be considered a source type.
+    * Note: as tags take part in type identify of the underlying struct literal, field tags are "inherited" in defined types based on existing types.
+      This may result in conflicting current behavior between those sources identified by tag and those identified by path/package/name.
+* A type which contains a known source type as a field should itself be a source type and associated field.  E.g., `type Wrapper struct { data core.Source }`
+* Current effort under the umbrella of *cross functional analysis* greatly refines inference of safety in scenarios that would currently produce a finding.
+
+## Reporting
+
+At present, reporting of Diagnostics is a generic `a source has reached a sink` message with source and sink positions included.
+As we iterate to include wider inference and definitions of sources, sinks, cross-function analysis, etc, the path between a source and a sink can become ambiguous.
+If much inference is targeted for a 1.0 release, reporting will need to be improved to remain meaningful.
+
+## Extensibility
+
+A user may have highly specific considerations for the definition of a source/sink/sanitizer and may wish to implement custom specification code that interacts directly with `ssa` / `types` packages.
+We should expose an extensible interface for this purpose.
```

Comment:
I think a reasonably easy way for us to do this would be to allow users to register a custom `analysis.Analyzer` as a `Requires` dependency of one of our analyzers, with a convention concerning the kind of results that it such an analyzer could produce. Our analyzer would then extract the results and use them in our analysis. E.g. if we wanted custom source detection, our `source` analyzer would extract the results from the user's custom analyzer and create sources from them.